dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/icon/DefaultIconService.java

```diff
@@ -67,33 +71,48 @@ public class DefaultIconService implements IconService {
   private static final String ICON_PATH = "SVGs";
 
   private final IconStore iconStore;
-
   private final FileResourceService fileResourceService;
-
   private final UserService userService;
 
+  private final Set<DefaultIcon> ignoredAfterFailure = ConcurrentHashMap.newKeySet();
+
   @Override
-  @Transactional
-  public void createDefaultIcon(DefaultIcon icon) {
-    if (iconStore.containsKeys(getKeysForVariants(icon))) return;
-
-    for (Icon i : createIconWithVariants(icon)) {
-      if (!iconStore.containsKeys(Set.of(i.getKey()))) {
-        String fileResourceId = CodeGenerator.generateUid();
-        Resource resource = getDefaultIconResource(i.getKey());
-        try {
-          FileResource fileResource =
-              fileResourceService.syncSaveFileResource(
-                  fileResourceId, MediaType.IMAGE_PNG, resource.getInputStream(), CUSTOM_ICON);
-
-          i.setFileResource(fileResource);
-
-          addIconInternal(i);
-        } catch (Exception ex) {
-          throw new RuntimeException(ex);
+  @Transactional(readOnly = true)
+  public List<Icon> findNonExistingDefaultIcons() {
+    Set<String> existing = Set.copyOf(iconStore.getAllKeys());
```

Comment:
We could directly return a `Set` from `getAllKeys` as this is what we need.

Reply:
I was back on forth on this one. I think it makes sense to expect a sorted response from a all values lookup. I just need fast lookup so I made it into a set later.